src/hotspot/share/opto/block.cpp

```diff
@@ -738,6 +730,41 @@ void PhaseCFG::remove_empty_blocks() {
       i--;
     }
   } // End of for all blocks
+
+  // Remove all blocks that are transitively unreachable after the
+  // NeverBranch-to-Goto conversion.
+  while (unreachable.size() > 0) {
+    Block* dead = unreachable.pop();
+    for (uint i = 0; i < number_of_blocks(); i++) {
+      Block* block = get_block(i);
+      if (block == dead) {
+        _blocks.remove(i);
+        _number_of_blocks--;
+        i--; // Ensure that we visit the block following the removed one.
+      }
+      if (block->_pre_order > dead->_pre_order) {
+        // Enforce contiguous pre-order indices (assumed by PhaseBlockLayout).
+        block->_pre_order--;
+      }
+      if (block->_rpo > dead->_rpo) {
+        // Enforce contiguous reverse post-order indices as well, for sanity.
+        block->_rpo--;
+      }
+    }
```

Comment:
Are you sure it will work for complex case? May be recalculate RPO and order after you removed dead blocks from graph.